### ISSUE TITLE
PE - How to avoid Cyclic Dependencies between 2 Classes

### ISSUE THREAD

**Issue by zhiqitan** [nus-cs2030/2021-s1#151]
Hi guys, I am facing issue of cyclic dependencies between Lecture class and Tutorial class when checking the clashWith, any tips for tackling this issue? especially the part of checking time. Thanks in advance ya!

**Comment by lkldev** [comment]
Hi, you might want to consider creating a method in the base class which is overridden by `Lecture` and `Tutorial` class. Then in the `Lecture` class, you do not have to check if it is `Tutorial` or `Lecture`, as its rules apply to all classes.  The `Tutorial` class will only need to check if the other class is `Tutorial` as its rules do not apply to `Lecture`. Hope this helps :)

**Comment by aaron-ljx** [comment]
If you're using instanceof in your code, you can only use instanceof Lecture in the Lecture class and likewise for the Tutorial class. You can't use instanceof Tutorial in the Lecuture class for example. That's how I avoided the cyclic dependencies.

**Comment by jadenwjh** [comment]
Why is cyclic dependency in this case bad? Since it doesn't cause StackOverflow error...

**Comment by lkldev** [comment]
@jadenwjh  I believe is due to testing? Like with cyclic dependency `Lecture` class can't be tested without `Tutorial` class and vice versa. It makes them very coupled and hard to reuse should the class grow bigger.

**Comment by zhiqitan** [comment]
> If you're using instanceof in your code, you can only use instanceof Lecture in the Lecture class and likewise for the Tutorial class. You can't use instanceof Tutorial in the Lecuture class for example. That's how I avoided the cyclic dependencies.

**Comment by zhiqitan** [comment]
> If you're using instanceof in your code, you can only use instanceof Lecture in the Lecture class and likewise for the Tutorial class. You can't use instanceof Tutorial in the Lecuture class for example. That's how I avoided the cyclic dependencies.

Thank you so much! It works for me to avoid the cyclic dependencies now!

**Comment by jadenwjh** [comment]
@lkldev i see... thanks!

btw, using the duration attribute to check if the object is lecture/tutorial works too if you defined the getter method in the parent class. The method will be provided by the parent class, so no cyclic dependency between Lecture and Tutorial.

**Comment by SkyBlaise99** [comment]
or you can just purely code the clash with inside Class.java instead

**Comment by ramapriyan912001** [comment]
Another way to go about this would be to use a common trait between the two subclasses that differs in value

In this case, the most obvious difference would be duration
Both have a duration, but this duration is different.

**Comment by zhiqitan** [comment]
> or you can just purely code the clash with inside Class.java instead

hi, if write clash with in Class.java, do you have any idea for how to check if it's a lecture or tutorial without using instanceof?

**Comment by lkldev** [comment]
Do you have a variable such as `ClassType` to indicate if the class is Lecture or Tutorial? If not you can just code the most generic `clashWith` that works for Lecture (i.e. no need to check if is lecture or tutorial) and in Tutorial, you override the `clashWith` method.

**Comment by SkyBlaise99** [comment]
class.java has info on duration of class which can distinguish between a lecture and a tutorial. A better approach I used is to make a flag isLec n isTut to better separate them up

**Comment by elizabethh07** [comment]
> If you're using instanceof in your code, you can only use instanceof Lecture in the Lecture class and likewise for the Tutorial class. You can't use instanceof Tutorial in the Lecuture class for example. That's how I avoided the cyclic dependencies.

@aaron-ljx Yes i used instanceof in my Classes.java and it caused cyclic dependencies. Do you know what's the reason? Thanks!

**Comment by SkyBlaise99** [comment]
I think it is because ur Class.java relies on checking if it is instanceof Lecture and your Lecture.java calls clashwith() in Class.java. So there is sort of like a cyclic dependency.

**Comment by zhiqitan** [comment]
> Do you have a variable such as `ClassType` to indicate if the class is Lecture or Tutorial? If not you can just code the most generic `clashWith` that works for Lecture (i.e. no need to check if is lecture or tutorial) and in Tutorial, you override the `clashWith` method.

hi, do you mind to give more hints on how to make the 'ClassType' to work?

**Comment by zhiqitan** [comment]
> class.java has info on duration of class which can distinguish between a lecture and a tutorial. A better approach I used is to make a flag isLec n isTut to better separate them up

hi, I don't have info on duration in class.java, how to do so if without using instanceof but still can distinguish the duration of class to be 1 hr or 2hr ar?

**Comment by SkyBlaise99** [comment]
You add a boolean isLecture in class.java?

**Comment by zhiqitan** [comment]
> You add a boolean isLecture in class.java?

how to check if its lecture at class.java?

**Comment by SkyBlaise99** [comment]
In lecture u call super right, that will parse info to class and you will parse the boolean flag to it also. Then in class, you can just use `if(isLecture)... `

**Comment by aaron-ljx** [comment]
> 
> 
> > If you're using instanceof in your code, you can only use instanceof Lecture in the Lecture class and likewise for the Tutorial class. You can't use instanceof Tutorial in the Lecture class for example. That's how I avoided the cyclic dependencies.
> 
> @aaron-ljx Yes i used instanceof in my Classes.java and it caused cyclic dependencies. Do you know what's the reason? Thanks!

I think it's something like this.

You used instanceof in Classes.java, which I assume is to check if something is a Lecture or Tutorial class.
So we can say Classes -> Lecture/Tutorial.
But since Lecture/Tutorial inherits from Classes, we also can say Lecture/Tutorial -> Classes.
This results in a cyclic dependency as they depend on each other to work. 
Thus, to break the dependency, you need to remove the instanceof in your Classes.java.

**Comment by zhiqitan** [comment]
> In lecture u call super right, that will parse info to class and you will parse the boolean flag to it also. Then in class, you can just use `if(isLecture)...`

do u mean super in the constructor? how to super at subclass's constructor when its not in the parameter? I get error of "call to super must be first statement in constructor".

**Comment by elizabethh07** [comment]
> > > If you're using instanceof in your code, you can only use instanceof Lecture in the Lecture class and likewise for the Tutorial class. You can't use instanceof Tutorial in the Lecture class for example. That's how I avoided the cyclic dependencies.
> > 
> > 
> > @aaron-ljx Yes i used instanceof in my Classes.java and it caused cyclic dependencies. Do you know what's the reason? Thanks!
> 
> I think it's something like this.
> 
> You used instanceof in Classes.java, which I assume is to check if something is a Lecture or Tutorial class.
> So we can say Classes -> Lecture/Tutorial.
> But since Lecture/Tutorial inherits from Classes, we also can say Lecture/Tutorial -> Classes.
> This results in a cyclic dependency as they depend on each other to work.
> Thus, to break the dependency, you need to remove the instanceof in your Classes.java.

@aaron-ljx okay thanks! but this error only arise in CodeCrunch? As vim does not give any errors.

**Comment by aaron-ljx** [comment]
> 
> 
> > > > If you're using instanceof in your code, you can only use instanceof Lecture in the Lecture class and likewise for the Tutorial class. You can't use instanceof Tutorial in the Lecture class for example. That's how I avoided the cyclic dependencies.
> > > 
> > > 
> > > @aaron-ljx Yes i used instanceof in my Classes.java and it caused cyclic dependencies. Do you know what's the reason? Thanks!
> > 
> > 
> > I think it's something like this.
> > You used instanceof in Classes.java, which I assume is to check if something is a Lecture or Tutorial class.
> > So we can say Classes -> Lecture/Tutorial.
> > But since Lecture/Tutorial inherits from Classes, we also can say Lecture/Tutorial -> Classes.
> > This results in a cyclic dependency as they depend on each other to work.
> > Thus, to break the dependency, you need to remove the instanceof in your Classes.java.
> 
> @aaron-ljx okay thanks! but this error only arise in CodeCrunch? As vim does not give any errors.

Yes, because cyclic dependency is a design problem. Bad designs such as cyclic dependency makes code hard to understand and maintain in the future, that's why CS2030 discourages cyclic dependency.

**Comment by SkyBlaise99** [comment]
@zhiqitan in your Lecture u should have
```
public Lecture(some parameter) {
    super(some parameter)
}
```

Just add a boolean flag or anything u prefer at the back
```
public Lecture(some parameter, boolean flag) {
    super(some parameter, flag);
}
```

Subsequently, modify for Class.java, Lecture.java, Tutorial.java
